File: spec/coffeescripts/jquery/jquerySpec.js
/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import 'jquery-migrate'

QUnit.module('JQuery', {
  beforeEach() {
    // Setup code that runs before each test
    $('#fixtures').append('<div id="test-jquery">Test JQuery</div>')
  },
  afterEach() {
    // Teardown code that runs after each test
    $('#fixtures').empty()
  },
})

QUnit.test('jquery can append to the document body', async function (assert) {
  // note: in jquery 1.9 or later, you can't append to the $(document), must be $(document.body) or other element
  const done = assert.async() // Get the async function from assert
  const $body = $(document.body)

  $("<div class='test-class'>hello world</div>").appendTo($body)
  const $appendedDiv = $body.find('.test-class')
  assert.equal($appendedDiv.text(), 'hello world')
  done()
})

QUnit.test('jquery can select stuff', async function (assert) {
  const done = assert.async() // Get the async function from assert
  const $div = $('#test-jquery')
  assert.equal($div.text(), 'Test JQuery')
  done()
})

File: spec/selenium/miscellaneous/jquery_spec.rb
# frozen_string_literal: true

#
# Copyright (C) 2011 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

require_relative "../common"

describe "jquery" do
  include_context "in-process server selenium tests"

  it "handles $.attr(method, post|delete|put|get) by adding a hidden input" do
    get("/login")
    expect(driver.execute_script("return $('form').attr('method', 'delete').attr('method')").downcase).to eq "post"
    expect(driver.execute_script("return $('form input[name=_method]').val()")).to eq "delete"
  end

  it "is able to handle ':hidden' and ':visible' pseudo-selector on document body" do
    get("/login")
    expect(driver.execute_script("return $(document.body).is(':visible')")).to be true
    expect(driver.execute_script("return $(document.body).is(':hidden')")).to be false
  end
end
